test-junit/jvm/src/main/scala/zio/test/junit/ZTestJUnitRunner.scala

```diff
@@ -0,0 +1,161 @@
+package zio.test.junit
+
+import org.junit.runner.manipulation.{ Filter, Filterable }
+import org.junit.runner.notification.{ Failure, RunNotifier }
+import org.junit.runner.{ Description, RunWith, Runner }
+import zio.ZIO.effectTotal
+import zio._
+import zio.clock.Clock
+import zio.test.FailureRenderer.FailureMessage.Message
+import zio.test.Spec.{ SpecCase, SuiteCase, TestCase }
+import zio.test.TestFailure.{ Assertion, Runtime }
+import zio.test.TestSuccess.{ Ignored, Succeeded }
+import zio.test._
+import com.github.ghik.silencer.silent
+
+class ZTestJUnitRunner(klass: Class[_]) extends Runner with Filterable with DefaultRuntime {
+  private val className = klass.getName.stripSuffix("$")
+
+  private lazy val spec: AbstractRunnableSpec = {
+    klass
+      .newInstance()
+      .asInstanceOf[AbstractRunnableSpec]
+  }
+
+  private var filter = Filter.ALL
+
+  lazy val getDescription: Description = {
```

Comment:
Is the issue here that JUnit needs a more event based interface and the way ZIO Test is currently structured we run all the tests at once and then render all the tests at once?

Currently we're unsafely running the specs once to get the descriptions and then a second time to run the tests, which could create issues if there are suite level effects such as with `provideManagedShared`. Would it be possible to move to more of a model where we execute all the specs and then report their results? I think this would allow us to avoid the double evaluation problem and reuse more of the existing functionality in ZIO Test.

Separately I do think we want to look into having an interface where it is easier to get notifications in real time about test events instead of waiting for the full output since that is a request we have gotten for ZIO Intellij as well but I think that is outside the scope of this PR.

Reply:
The leaf tests are not actually evaluated, just the `specs` effects in the suites.
I can actually return an empty `Descriprion` and just report test events as the happen.
(I'm actually reporting start/finish/fail events asynchronously already - look at `instrumentSpec` method).
In this case however, there is no way to report the nested structure of the Spec, the results just come out flat. 
The best I can do is to include the path in the label of each leaf test.

Reply:
The problem is the `specs` effects in the suites could have real logic in them. They might create a Kafka client for example and then dispose of it.

Can we have the test executor execute the specs so we get one set of executed specs that don't have any environmental dependency and then traverse that to report everything?

Reply:
Well, I don't think having the spec executed in `getDescription` instead of `run` will be following the JUnit Runner contract (if that's what you mean). IDEs and build tools won't expect this. And also you will loos the live reporting of the test as they run.
The only other option I see to prevent double evaluation is to have empty description and and accept flat spec structure being reported from inside the `run` method.

Reply:
Yes, we're running into an issue where specs in JUnit are static (there are just a certain number of tests sitting out there) versus tests in ZIO Test are dynamic (creating tests in a spec is itself an effect).

I see two options:
1. `getDescription` evaluates suite level effects and returns the full structure of the spec. The advantage is you get a description up front with all the suite and test labels. The disadvantage is that you run all suite level effects, like acquiring and releasing shared managed values, when you get the description, which to your point is probably not the contract for getting a description, and then you have to evaluate them again when calling `run`, which is inefficient.
2. `getDescription` just returns the label of the top suite or test and doesn't evaluate anything. I think we could still include information on the full ancestry of tests in events we report for each test by either traversing the spec and recursively passing down a list of ancestors or keeping track of it in a `Ref`. The disadvantage is that calling that `getDescription` method won't be that informative, but it would accurately reflect that we don't have static descriptions beyond the top level label before running the spec and would avoid the double evaluation issue.

I think the second approach is preferable. What do you think?

Reply:
I'm good with 2, but I couldn't find a way to report the full ancestry of the leaves being executed.
I can the report the direct parent though. So what I can do is concatenate the full path in the parent name, so in IntelliJ it will look something like this:
```
SomeSpecWithJUnit
   + suite 1:suite 2
      - test1
      - test2 
   + suite 3:suite 4
      - test3
      - test4 
```

Reply:
Unfortunately, I just discovered that while this works with IDEA JUnit runner, returning empty or fake description doesn't work with Bazel. It seems to expect tests to have accurate descriptions. This may also be the case for Gradle and Maven.

Reply:
Sorry I missed this. Can you say a little more about what you mean by "accurate descriptions"? Trying to understand if there is something we can do about that or that effectively means that Option 2 above is not possible.

Reply:
Well, I can describe what is happening in our Bazel environment with Intellj with our customized version of Bazel Intellij plugin, when the runner returns empty description.
When running all tests under target (package) Bazel output reflects the tests run and failures, but IDEA is showing success and doesn't reflect tests run.
When running a single test using a filter like `--test_filter=com.wixpress.infra.zio.test.SimpleFailingTest#` we get  `No tests found matching RegEx[com.wixpress.infra.zio.test.SimpleFailingTest#]`.
It seems that bazel test filter for JUnit expects the class name to appear in the descriptions of leaf tests.

Reply:
So you think that if we got the full ancestry to the leaves it would?

Reply:
I've been testing it in our project and if the `getDescription` returns the hierarchical structure of our spec with the spec class name included in the leaves everything works in our environment.
By class name included in the leaves I mean `Description.createTestDescription(className, label)` 

![image](https://user-images.githubusercontent.com/1577551/71830555-a7bfdb00-30af-11ea-91de-e7a0e0219f3f.png)

Reply:
Okay, so how about we do this?

Create a new simple algebraic data type for a description:

```scala
sealed trait SpecDescription extends Product with Serializable {
  // Converts to JUnit's representation of a description
  def toDescription: Description = ???
}
final case class TestDescription(label: String) extends SpecDescription
final case class SuiteDescription(label: String, specs: Vector[SpecDescription]) extends SpecDescription
```

Create an `AtomicReference` to store the description so it is safe for concurrent access, initially populated with just the top level spec name:

```scala
private val description: AtomicReference[SpecDescription] = ???
```

Then `getDescription` becomes:
```scala
`def getDescription = description.get.toDescription
```

Then the final step is in `traverse` we add an additional parameter `ancestors` and add the current label to `ancestors` each time we recurse. This will give us the information we need to update the description as we go.

I think this is all very doable. The upshot though is that `getDescription` will return a different description depending on when it is evaluated (e.g. if it evaluated before any tests are run it will just show the top level label, as suites and tests are created theyl be added to the description). So we need to make sure that works. Otherwise I think we will have to fall back to Option 1 above.

Thoughts?

Reply:
Well at least for bazel, `getDiscription` is called before running the test, and filtering considers the description returned, this will not work, at least not for filtering by class name...
As far as I can tell, this is the contract - `getDescription` is called before `run` and is not expected to change if you called again.
It looks like Option 1 is the way to go, unfortunately.
Perhaps there could be a way to separate the suite effect into "dry run" mode - which only generates the structure and "actual run" mode that also performs setup and cleanup effects and such? 
What if SuiteSpec.specs was `Mode => ZIO[R, E, Vector[A]]` where Mode is either `Run` or `Describe`?